### PR TITLE
add node shutdown taint

### LINKED ISSUE #58635: Propose to taint node "shutdown" condition
Today there is only one node condition to represent whether kubelet is healthy (updating node status). Node controller will mark node as "NotReady" after kubelet stops updating for a while. The reason for kubelet not responding could be 
1. Network disconnect
2. Node is shun down (stopped) or paused
3. Kubelet is killed (could not response)

## Problem
In many cases, when a node becomes "NotReady", pods will be evicted, and new pods will be started on other nodes (failover for high availability). However, if node is shut down, the volumes attached to the nodes are still attached. Volume controller should try to detach volumes when pods are killed from the old node. But there is a safety check before detaching to see whether volume is still mounted. This check is done by a VolumeInUse field in node status which is updated by kubelet. In case of node shut down, this field will not be updated by kubelet so that the safety check will fail and assume the volume cannot be detached because of existing mounts. The new pods will be pending for a while because volume cannot be attached to the other node. Although the reality is that volume can be detached in this situation. (Note: some cloud provider will treat stopped node as not exist in cloud provider so that node controller will delete the node API object. But still detach will not happen because of the mount safety check. Ref https://github.com/kubernetes/kubernetes/issues/46442)

## Proposal
We propose to taint the node with "Stopped" condition. When kubelet stops updating, node controller checks cloud provider and see whether the node is in "stopped" state (Note: different cloud provider might use different term to represent this state, e.g., terminated). When volume controller gets a node update/delete event, it can check the node taint spec and mark the volumes attached to that node are safe to detach (volumes can be be removed from the list of VolumeInUse (meaning the volumes are mounted). When detach is issued, the operation can be proceeded without delay.

To determine the "stopped" condition, I lists a few cloud provider's explanation of stopping an instance.

### GCE

- Stopping an instance causes Compute Engine to send the ACPI Power Off signal to the instance. Modern guest operating systems are configured to perform a clean shutdown before powering off in response to the power off signal. Compute Engine waits a short time for the guest to finish shutting down and then transitions the instance to the TERMINATED state.

- You can stop an instance temporarily so you can come back to it at a later time. A stopped instance does not incur charges, but all of the resources that are attached to the instance will still be charged. Alternatively, if you are done using an instance, delete the instance and its resources to stop incurring charges.

- You can permanently delete an instance to remove the instance and the associated resources from your project. If the instance is part of an instance group, the group might try to recreate the instance to maintain a certain group size.

### AWS

- You can only stop an Amazon EBS-backed instance. The instance performs a normal shutdown and stops running; its status changes to stopping and then stopped. Any Amazon EBS volumes remain attached to the instance, and their data persists. In most cases, the instance is migrated to a new underlying host computer when it's started. 

- You can delete your instance when you no longer need it. This is referred to as terminating your instance. When an instance terminates, the data on any instance store volumes associated with that instance is deleted. If your instance is in an Auto Scaling group, the Auto Scaling service marks the stopped instance as unhealthy, and may terminate it and launch a replacement instance.

### OpenStack

- Admins can pause and unpause a Nova compute instance. When an instance is paused, the entire state of the instance is kept in RAM. Pausing an instance will disable access to that instance, but won't free up any of its resources. Another option is to suspend, and then resume, an instance. Like paused OpenStack instances, a suspended instance keeps its current state, but it is written to storage. 

- A third option is to shelve OpenStack instances.A shelved instance is actually shut down, which is not the case for suspended or paused instances. If admins decide they no longer need a shelved instance, they can remove it, which ensures that it doesn't maintain any hypervisor-level resources in use.

- The last option is to stop an instance in Nova, which will disconnect all of its associated resources. This means admins can't restore a stopped instance to its previous state. This option is only useful for OpenStack instances that an organization no longer needs. In all other cases, admins should shelve, suspend or pause the instance.

So basically a "stopped" node means that it is shut down like an operating system shut down situation. All mounts are gone when machine is shut down.

cc @yujuhong @dchen1107 @Random-Liu

### PR DESCRIPTION
**Pull request by zetaab**
**What this PR does / why we need it**: we need node stopped taint in order to detach volumes immediately without waiting timeout. More info in issue ticket #58635 

**Which issue(s) this PR fixes** 
Fixes #58635

**Special notes for your reviewer**: this was reverted, original PR https://github.com/kubernetes/kubernetes/pull/59323 Hopefully now bugs are fixed. However, I will execute more tests manually today.

**Release note**:
```release-note
Add taint on node shutdown to allow immediate volume detach
```

### CODE REVIEW

**Comment by zetaab** [comment]
I have executed following tests manually:

tests:

without cloudprovider fix added (another PR open)
- nodecontroller shutdown machine (should delete node from cluster)
- nodecontroller delete machine from cloud (should delete node from cluster) [last time this use-case contained bug. It did not delete the node from cluster]
- cloud controller shutdown machine (should delete node from cluster) 
- cloud controller delete machine from cloud (should delete node from cluster)

with cloudprovider fix added (openstack PR #59931)

- nodecontroller shutdown machine (should add taint and not delete node)
- nodecontroller delete machine from cloud (should delete node from cluster)
- cloud controller shutdown machine (should add taint and not delete node)
- cloud controller delete machine from cloud (should delete node from cluster)

Now all of these seems to work as should.

**Comment by zetaab** [comment]
cc @jingxu97 @jhorwit2 @gmarek @yastij @dims could you guys review

**Comment by deads2k** [comment]
@jhorwit2 you lgtm'ed the original.  There's a separate commit with the diff.  Let's try to close this off during slush

**Comment by jhorwit2** [comment]
@zetaab could you please provide some context to the bugs that were fixed? This seems largely the same as the previous PR minus 1 if statement from what I see.

**Comment by zetaab** [comment]
@jhorwit2 @deads2k last time there was bug: if instance was deleted from cloudprovider, this code failed to this row https://github.com/kubernetes/kubernetes/pull/60009/commits/2eef27c40c13a50d2f8aa21ca0a0b72b92a467b0#diff-da876211e68ffbdebf956f69166e6791L197 it returned error because the instance did not exist. This lead to situation where we did not have instances in cloudprovider anymore but nodes existed in cluster. In correct situation those nodes should be deleted. 

Like you can see, now I am not trying to fetch providerID from cloudprovider. Instead I am using node.Spec providerid because we know that already (one api call less also).

**Comment by yastij** [comment]
@zetaab - I’ll do a pass tomorrow

File: pkg/controller/cloud/node_controller.go
Comment (jhorwit2) [inline pkg/controller/cloud/node_controller.go]: `}` should be on the next line.

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go
Comment (jhorwit2) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: We should put this in a common place now that it's used in multiple packages

File: pkg/controller/util/node/controller_utils.go
Comment (jhorwit2) [inline pkg/controller/util/node/controller_utils.go]: this should take a context

File: pkg/controller/cloud/node_controller.go
Comment (jhorwit2) [inline pkg/controller/cloud/node_controller.go]: Could we reuse the logic in this util here? `pkg/controller/util/node/controller_utils.go`. I just noticed a bug in `ExistsInCloudProvider` because we didn't reuse the same code across controllers I need to open a PR for.

File: pkg/controller/util/node/controller_utils.go
Comment (jhorwit2) [inline pkg/controller/util/node/controller_utils.go]: I think you should add the not implemented check here and return `false, nil` if it's not implemented to reduce the error handling in the controllers.

**Review by jhorwit2 — CHANGES_REQUESTED** [review]
Why is the logic the exact same in two controllers that appear to be updating on the same interval? Can't we just have this in the node lifecycle controller?

**Comment by jhorwit2** [comment]
> Why is the logic the exact same in two controllers that appear to be updating on the same interval? Can't we just have this in the node lifecycle controller?

You can ignore that. I got filled in on this in slack. The eventual plan once we're done running cloud things in the KCM is to remove the duplicate logic.

**Comment by zetaab** [comment]
@jhorwit2 bow ok?

**Review by yastij — COMMENTED** [review]
This LGTM overall, @jhorwit2 to you for confirmation

File: pkg/controller/cloud/node_controller.go
Comment (jhorwit2) [inline pkg/controller/cloud/node_controller.go]: The node isn't fine if we're here, it's shutdown.

**Comment by zetaab** [comment]
@jhorwit2 fixd

**Comment by jhorwit2** [comment]
@zetaab this lgtm. You need to get approvals though from the relevant areas tho.

**Comment by BenTheElder** [comment]
/test pull-kubernetes-unit

Verfiy is blocked on https://github.com/kubernetes/kubernetes/pull/60442

File: pkg/controller/controller_utils.go
Comment (jingxu97) [inline pkg/controller/controller_utils.go]: Maybe ShutdownTaint to be consistent

**Comment by jingxu97** [comment]
I have some question related to the node name/provider id. See issue https://github.com/kubernetes/kubernetes/issues/60466

**Comment by jingxu97** [comment]
cc @yujuhong @Random-Liu

**Comment by zetaab** [comment]
@bsalamat @k82cn @deads2k @jsafrane @dims could you check this. I need approvals from all areas.

**Comment by zetaab** [comment]
@mikedanese could you check this

**Comment by yastij** [comment]
cc @dims @wlan0 can you approve this one guys ? thanks

**Review by feiskyer — APPROVED** [review]
LGTM. Is this PR targeted for v1.10?

**Comment by zetaab** [comment]
@feiskyer https://github.com/kubernetes/kubernetes/issues/58635#issuecomment-362743865 according this yes the target is 1.10

**Comment by zetaab** [comment]
@jhorwit2 can you or someone else add /lgtm

**Comment by jhorwit2** [comment]
@zetaab since this is in code freeze we would need someone to add the approved for milestone label to get this into 1.10. @dims can you help with that?

**Comment by dims** [comment]
@zetaab @jhorwit2 i don't have karma to add milestone. i think we are already in the phase where we have to ask for exceptions (example https://groups.google.com/d/msg/kubernetes-milestone-burndown/pwI3QwuCYL4/ugg9UdhWAQAJ)

**Comment by yastij** [comment]
@zetaab - can you rebase this ?

**Comment by yastij** [comment]
@jhorwit2 @dims - can you apply lgtm ?

**Comment by dims** [comment]
@yastij @zetaab will need to remove the extra commit and squash the other two before we can proceed

**Comment by yastij** [comment]
@zetaab - can you squash commits ?

**Review by yastij — APPROVED** [review]
lgtm @dims

File: pkg/controller/controller_utils.go
Comment (k82cn) [inline pkg/controller/controller_utils.go]: should be in `pkg/controller/nodelifecycle/node_lifecycle_controller.go`

**Comment by k82cn** [comment]
@zetaab , can you also give some meaningful commit comments, not just `fix`?

File: pkg/controller/controller_utils.go
Comment (zetaab) [reply pkg/controller/controller_utils.go]: I do not fully agree with this, because then we need import nodelifecycle package to https://github.com/zetaab/kubernetes/blob/b900bf3ed391efd774237d491bdf50780057941e/pkg/controller/cloud/node_controller.go or then make same variable there? Which one is better 1) import nodelifecycle package to cloudcontroller (the worst option imo) 2) make twice same variable 3) keep it as is (in controller_utils)

Comment (yastij) [reply pkg/controller/controller_utils.go]: @k82cn - I’d prefer keeping this here for now, I’ll open a dedicated PR to move all the taints to the same place.

Comment (k82cn) [reply pkg/controller/controller_utils.go]: that's ok to me :)

**Comment by yastij** [comment]
@zetaab - can you update your message commit ?

**Comment by dims** [comment]
@zetaab this ready to merge?

**Comment by zetaab** [comment]
@dims yes

**Pull request merged**
